Fix some bugs. Also add a driver app that uses odmpy libby to automatically get all your checkouts.

wtanksleyjr opened the pull request:
Had way too much fun with this one. Some details:

- ~Documentation is pending.~ done.
- Random bugs fixed: parts->chapters map generates correctly, using the play toggle waits a bit longer, and so on.
- Allow pinning the Dockerfile to a specific build of selenium; I discovered that they randomly change existing containers so docker compose's built in autobuild would randomly spaz out and download a ton needlessly.
- Selenium's containers change constantly; added a build tool, `libbyload.py --build-only` which uses the newly provided docker-compose.yml file and the slightly modified Dockerfile to pin and build in a way that can easily be updated (if you run `libbyload --build-only` again, it'll only rebuild once a day unless you actually changed a file Docker needs like requirements.txt or entrypoint.sh). Some setup required for libbyload even if you just use build-only: either pass it your downloads dir using the `-d` option, or set an environment variable AUDIOBOOK_FOLDER to your download folder and AUDIOBOOKS_TMP to your work in progress folder.
- There's a tmp folder now, we were using the source code folder, yikes.

wtanksleyjr commented:
DUDE I found the craziest Overdrive bug. One of the books I checked out downloaded an absolutely MESSED UP part31, it showed as having a slightly short size which made odmpy-ng spend HOURS trying to find a last part that didn't exist; but audiobookshelf saw it as corrupted and loaded only the other parts, making it obviously too short.

ANYHOW: my fix attempt is to add a check for the MP3 reader's property "shoddy", which may help for some kinds of corruption. There's no real solution though, this file downloaded corrupted every time.

kernalbin commented:
Probably just a logging error, but I tried downloading a book, it said it downloaded 4 parts, then it said the book contained 5 parts and then when converting to m4b it only had 4. The end file was the correct length and was complete though.

```
Accessing The Fearless Mind, ID: 10571108
Getting chapters
Got 46 chapters
Final book should be ~4:57:22 in length.
Getting chapter:part structure.
Building structure chapter:part (47 chapters): 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46
Found end of book at end of chapter 46, part 4, at 04:57:21
Got part->chapter structure: ['01->0 (00:00:00)', '02->12 (01:18:22)', '03->20 (02:30:58)', '04->34 (03:48:51)']
Getting files
Checking for resumable parts.
Downloading part 1
01:13:46 / 04:57:22 - 4426/17842 sec  -  24.81%
Downloading part 2
02:30:57 / 04:57:22 - 9057/17842 sec  -  50.76%
Downloading part 3
03:40:19 / 04:57:22 - 13219/17842 sec  -  74.09%
Downloading part 4
04:57:21 / 04:57:22 - 17841/17842 sec  -  99.99%
Downloaded complete audio
Book contained 5 part(s)
Downloaded cover
Downloaded json metadata
Provided audiobookshelf metadata
Cleaned up json metadata
4 MP3 file(s) found
```

I also may just be using it wrong. But I kept getting an error about /tmp/libby.json not existing until I just created an empty file there.

kernalbin commented:
Aside from that. The features added here are awesome.

Though, I'm a bit uncertain about how accessible the odmpy integration will be for most people. The functionality is great, but I wonder if the overall setup cost, installing and configuring both odmpy and ng, would be a barrier to many potential users. 

I don't have concrete suggestions yet on how to streamline things, I'll be playing with it some more and see if I can't come up with anything. And maybe I'm overestimating the friction I'm imagining this creates.

Would love to hear your thoughts. What do you think of the setup? Whats your endgame here, what does the ideal library downloader look like to you?

wtanksleyjr commented:
Ha, oh no, that log message displays after the increment now. It's just a logging error.

kernalbin commented:
When you publish your Libby integrated version lmk. I want to link to it. (:

wtanksleyjr commented:
Finally wrote the docs, and while I was doing it I realized what I REALLY needed to add - a "run" tool to manage the absurd command line complexity with UID and paths and so on. Check it out (in the docker-compose section of the docs), it's WAY nicer.

wtanksleyjr commented:
I'll definitely let you know - I have it working again, but there's a few things I want to do before I make a "release", like basing it against my main branch and so on.

Also - I think I should add a table to the docs discussing the options. I added several new ones, but it also would be nice to explain the old ones.

One thought ... back in another PR, I made the source code read-only, and it turns out you were using that to write tempfiles (yikes). So I moved the tmp directory to the only sensible place, downloads (I wouldn't want to write onto the config folder). But this has its own challenges ... if the user placed the download folder directly into an audiobook manager (like audiobookshelf), it'll possibly see a book added and then removed, which could be annoying (usually book managers will throw an error). So probably we should add an external tmp folder as well. I didn't want to do that to people using the regular docker command line (it adds more and more to the length of each command), but it's really not that bad for docker-compose, we only take one variable right now so adding one more isn't that bad.

What do you think?

wtanksleyjr commented on .dockerignore:
```diff
@@ -0,0 +1,4 @@
+*
+!entrypoint.sh
+!requirements.txt
```
Reduce the number of rebuilds; because almost all of the source is not in the Docker anymore, only depend on files that are actually needed.

wtanksleyjr commented:
OK, documentation looks better. IOU a temp directory option next.

kernalbin commented:
You don't owe me anything haha. But I appreciate it.

wtanksleyjr commented:
I've spent too much time writing documentation, and I'm still not really finished with the docs, but here's the program that drives odmpy-ng using these new options. I also added a patch to make the old odmpy give me a bit more info, and the result is that it generates the library entries in the odmpy-ng config.json file.

I've been using this for a couple of weeks, seems reasonably reliable.

https://github.com/wtanksleyjr/odmload

As before it's just a single python file.
